### PR TITLE
Fix helm repo index with encoded URLs

### PR DESCRIPTION
**Pull request by sathieu**
**What this PR does / why we need it**:

While fixing #9764, and #9821, I found another usage of `url.Path` without `url.RawPath`.

I have not (yet) an end-user case, but the code is only used for `helm repo index`.

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by sathieu** [comment]
@cndoit18 This one is an edge-case, but it's still a bug I think. Please review.

**Review by cndoit18 — APPROVED** [review]

**Comment by sathieu** [comment]
@cndoit18 You approved those changes. What's next?

**Comment by sathieu** [comment]
@hickeyma Please review :pray:.

**Comment by sathieu** [comment]
@cndoit18 @bacongobbler Please review again. I've rebased and fixed conflicts.

**Comment by sathieu** [comment]
@bacongobbler Anything missing to merge this PR?

**Comment by sathieu** [comment]
This PR is now 10 months old. Could it be added to the [3.9 milestone](https://github.com/helm/helm/milestone/116) to ensure it is reviewed? :pray:

**Review by yxxhero — APPROVED** [review]
LGTM

File: internal/urlutil/urlutil_test.go [hunk 1]
```diff
@@ -28,6 +28,9 @@ func TestURLJoin(t *testing.T) {
 		{name: "URL, two paths", url: "http://example.com", paths: []string{"hello", "world"}, expect: "http://example.com/hello/world"},
 		{name: "URL, no paths", url: "http://example.com", paths: []string{}, expect: "http://example.com"},
 		{name: "basepath, two paths", url: "../example.com", paths: []string{"hello", "world"}, expect: "../example.com/hello/world"},
+		{name: "encoded parts", url: "../example.com", paths: []string{"hello", "world"}, expect: "../example.com/hello/world"},
```
Comment (joejulian) [inline internal/urlutil/urlutil_test.go:31]: Am I missing something obvious, or are these two tests cases exactly the same?

File: internal/urlutil/urlutil_test.go [hunk 2]
```diff
@@ -28,6 +28,9 @@ func TestURLJoin(t *testing.T) {
...
+		{name: "Long URL with encoded path, non-encoded paths", url: "http://example.com/but%2ffirst", paths: []string{"part-1", "part-2"}, expect: "http://example.com/but%2ffirst/part-1/part-2"},
+		{name: "Long URL with encoded path, encoded paths", url: "http://example.com/but%2ffirst", paths: []string{"part%25-1", "part%25-2"}, expect: "http://example.com/but%2ffirst/part%25-1/part%25-2"},
```
Comment (joejulian) [inline internal/urlutil/urlutil_test.go:33]: How would we expect a file path with url encoding to work? ie:
```
{name: "file with url encoding", url: "../example.com", paths: []string{"hello%3A", "world%2F", expect ... 
```

**Review by joejulian — CHANGES_REQUESTED** [review]
Looks good for the most part.

Please review the test cases

File: internal/urlutil/urlutil_test.go (hunk 2 above)
Comment (sathieu) [reply internal/urlutil/urlutil_test.go:33]: I think that encoding should be preserved. I've added a test for this.

File: internal/urlutil/urlutil_test.go (hunk 1 above)
Comment (sathieu) [reply internal/urlutil/urlutil_test.go:31]: Good catch, I've added encoded parts.

**Comment by sathieu** [comment]
Thanks @joejulian. I've rebased and addressed your comments. Back to you :ping_pong:!
